website/blog/2024-08-14-otel-context-propagation-in-ferretdb.md

```diff
@@ -0,0 +1,160 @@
+allowing you to link different parts of a distributed request together.
+The most common approach to implement context propagation is by using [trace context](https://www.w3.org/TR/trace-context/)
+headers to pass this information between services, but this method is not always feasible.
+
+Most databases don't support such headers or other native ways to pass tracing-related context.
+Approaches like [SQLCommenter](https://google.github.io/sqlcommenter/) have been developed to bridge this gap,
+enabling the connection between current trace data and database queries.
+In these cases, trace information is injected at the ORM level and passed to the database via SQL comments.
+Database operators can then enable query logs to link specific queries with the corresponding application requests.
+
+But is it possible to achieve something similar with MongoDB's query language?
+At FerretDB, we believe that enabling the linking of a query's lifecycle with the caller's response can help developers make
+their applications more predictable and reliable.
+
+## FerretDB's approach to context propagation
+
+Let's consider a SQLCommenter-like approach for MongoDB.
+We can leverage MongoDB's existing features to attach metadata to queries.
+One such feature is the comment field, which allows additional information to be included in MongoDB queries.
+This field doesn't influence query logic but can be used to provide valuable context.
+
+At FerretDB, we want to empower users to pass such context.
+We decided to utilize the `comment` field to do it.
+We parse the content of the `comment` field, and if it's a JSON document with a `ferretDB` key, we check for tracing data.
+If the tracing data is present, FerretDB sets the parent's context when a span for the current operation is created.
+
+An example of a comment with tracing data could look like this:
+
+```json
+{
+  "ferretDB": {
+    "traceID": "1234567890abcdef1234567890abcdef",
+    "spanID": "fedcba9876543210"
+  }
+}
+```
+
+The strings `traceID` and `spanID` are hex-encoded strings that represent the corresponding trace and span identifiers.
+
+Using this approach, it's relatively easy to visualize client's requests to FerretDB showing them as part of the original client trace.
+
+However, there are some limitations.
+Not all operations support the comment field.
+For instance, operations like `insert` or `listCollections` do not support it.
+
+Another challenge with the `comment` field is its handling in some MongoDB drivers, where it is typed as a string.
+This means that, in principle, any string can be passed.
+In our scenario, where we allow FerretDB to receive and parse
+tracing data through the `comment` field, the JSON must be correctly encoded by the driver and then decoded by FerretDB.
+
+A more robust solution would involve having a dedicated field for passing request-related context,
+and it would be ideal to establish a standard for such fields.
```

Comment:
https://github.com/FerretDB/FerretDB/pull/4512#pullrequestreview-2235093373 might be somewhere there